src/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/CollectionServices.cs

```diff
@@ -0,0 +1,298 @@
+                    return source;
+                }
+            }
+            
+            // last resort - nothing is null, need to append
+            source.AddRange(second);
+            return source;
+            
+        }
+
+        
+        private static List<T> FastAppendToListAllowNulls<T>(
+                    this List<T> source, T value)
+        {
+            if (source == null)
+            {
+                source = new List<T>();
+            }
+            source.Add(value);
+
+            return source;
+        }
+
+        public static List<T> FastAppendToListAllowNulls<T>(
+                        this List<T> source, T value,
+                        IEnumerable<T> second)
+        {
+            if (second == null)
+            {
+                source = source.FastAppendToListAllowNulls(value);
+            }
+            else
+            {
+                source = source.FastAppendToListAllowNulls(second);
+            }
+            return source;
+        }
+ 
+        public static void ForEach<T>(this IEnumerable<T> source, Action<T> action)
+        {
+            foreach(T t in source)
+            {
+                action.Invoke(t);
+            }
+        }
+
+        public static EnumerableCardinality GetCardinality<T>(this IEnumerable<T> source)
+        {
+            Assumes.NotNull(source);
+
+            // Cast to ICollection instead of ICollection<T> for performance reasons.
+            ICollection collection = source as ICollection;
+            if (collection != null)
+            {
+                switch (collection.Count)
+                {
+                    case 0:
+                        return EnumerableCardinality.Zero;
+
+                    case 1:
+                        return EnumerableCardinality.One;
+
+                    default:
+                        return EnumerableCardinality.TwoOrMore;
+                }
+            }
+
+            using (var enumerator = source.GetEnumerator())
+            {
+                if (!enumerator.MoveNext())
+                {
+                    return EnumerableCardinality.Zero;
+                }
+
+                if (!enumerator.MoveNext())
+                {
+                    return EnumerableCardinality.One;
+                }
+
+                return EnumerableCardinality.TwoOrMore;
+            }
+        }
+
+        public static bool FastAny<T>(this IEnumerable<T> source)
+        {
+            // Enumerable.Any<T> underneath doesn't cast to ICollection, 
+            // like it does with many of the other LINQ methods.
+            // Below is significantly (4x) when mainly working with ICollection
+            // sources and a little slower if working with mainly IEnumerable<T> 
+            // sources.
+
+            // Cast to ICollection instead of ICollection<T> for performance reasons.
+            ICollection collection = source as ICollection;
+            if (collection != null)
```

Comment:
all this can be simplified to: `return collection != null ? collection.Count > 0 : source.Any();`